lib/runtime-core/src/vm.rs

```diff
@@ -401,6 +399,17 @@ impl Ctx {
         }
     }
 
+    /// Get mutable access to the user defined data
+    /// field as the type, `T`.
+    ///
+    /// # Safety
+    ///
+    /// This function must be called with the same type, `T`, that the `data`
+    /// was initialized with.
+    pub fn data_mut<T>(&self, data_key: &str) -> &mut T {
```

Comment:
This function should be marked `unsafe` if it can lead to memory issues through safe Rust, otherwise the API is unsound!

Additionally, it seems like there's another implicit guarantee here that needs to be documented (or the API changed), the returned value does not include the lock that's preventing the hash map from being accessed.  This means that because we're passing in a `&self`, we can, in safe Rust, get 2 `&mut` to the same entry in the hashmap (this is also a problem from multiple threads)